data/model/autoprune.py

```diff
@@ -218,3 +233,211 @@ def namespace_has_autoprune_task(namespace_id):
+        )
+        return db_for_update(query, skip_locked=SKIP_LOCKED).get()
+    except AutoPruneTaskStatus.DoesNotExist:
+        return []
+
+
+def fetch_autoprune_task_by_namespace_id(namespace_id):
+    try:
+        return (
+            AutoPruneTaskStatus.select().where(AutoPruneTaskStatus.namespace == namespace_id).get()
+        )
+    except AutoPruneTaskStatus.DoesNotExist:
+        return None
+
+
+def delete_autoprune_task(task):
+    try:
+        task.delete_instance()
+    except AutoPruneTaskStatus.DoesNotExist:
+        return None
+    except Exception as err:
+        raise Exception(
+            f"Error deleting autoprune task for namespace id: {task.namespace_id} with error as: {str(err)}"
+        )
+
+
+def prune_repo_by_number_of_tags(repo, policy_config, namespace, tag_page_limit):
+    """
+    Prunes tags in the given repository based on the number of tags specified in the policy config.
+    """
+
+    policy_method = policy_config.get("method", None)
+
+    if policy_method != AutoPruneMethod.NUMBER_OF_TAGS.value:
+        raise InvalidNamespaceAutoPruneMethod(
+            f"Expected prune method type {AutoPruneMethod.NUMBER_OF_TAGS.value} but got {policy_method}"
+        )
+
+    assert_valid_namespace_autoprune_policy(policy_config)
+
+    while True:
+        tags = oci.tag.fetch_paginated_autoprune_repo_tags_by_number(
+            repo.id, int(policy_config["value"]), tag_page_limit
+        )
+        if len(tags) == 0:
+            break
+
+        for tag in tags:
+            try:
+                oci.tag.delete_tag(repo.id, tag.name)
+                log.log_action(
+                    "autoprune_tag_delete",
+                    namespace.username,
+                    repository=repo,
+                    metadata={
+                        "performer": "autoprune worker",
+                        "namespace": namespace.username,
+                        "repo": repo.name,
+                        "tag": tag.name,
+                    },
+                )
+            except Exception as err:
+                raise Exception(
+                    f"Error deleting tag with name: {tag.name} with repository id: {repo.id} with error as: {str(err)}"
+                )
+
+
+def prune_repo_by_creation_date(repo, policy_config, namespace, tag_page_limit=100):
+    """
+    Prunes tags in the given repository based on the creation date specified in the policy config,
+    where the value is a valid timedelta string. (e.g. 1d, 1w, 1m, 1y)
+    """
+
+    policy_method = policy_config.get("method", None)
+
+    if policy_method != AutoPruneMethod.CREATION_DATE.value:
+        raise InvalidNamespaceAutoPruneMethod(
+            f"Expected prune method type {AutoPruneMethod.CREATION_DATE.value} but got {policy_method}"
+        )
+
+    assert_valid_namespace_autoprune_policy(policy_config)
+
+    time_ms = int(convert_to_timedelta(policy_config["value"]).total_seconds() * 1000)
+    while True:
```

Comment:
Likewise, should this be wrapped in a transaction, in order to avoid possible conflict with concurrent tags operations

Reply:
yeah, we probably shouldn't do this, continuing the conversation in another thread